### ISSUE TITLE
Feature Request: Apple Silicone Neural Engine - Core ML model package format support

### ISSUE THREAD

**Issue by qdrddr** [AugustDev/enchanted#101]
**Description**

Please consider adding Core ML model package format support to utilize Apple Silicone Nural Engine + GPU.

**Success Criteria**
Utilize both ANE & GPU, not just GPU on Apple Silicon

**Additional Context**

List of Core ML package format models
https://github.com/likedan/Awesome-CoreML-Models

Work in progress on CoreML implementation for [whisper.cpp]. They see x3 performance improvements for some models. (https://github.com/ggerganov/whisper.cpp/discussions/548) you might be interested in. 

You might also be interested in another implementation [Swift Transformers](https://huggingface.co/blog/swift-coreml-llm). Example of CoreML application
https://github.com/huggingface/swift-chat

Core ML is a framework that can redistribute workload across CPU, GPU & Nural Engine (ANE). ANE is available on all modern Apple Devices: iPhones & Macs (A14 or newer and M1 or newer). Ideally, we want to run LLMs on ANE only as it has optimizations for running ML tasks compared to GPU. Apple claims ["deploying your Transformer models on Apple devices with an A14 or newer and M1 or newer chip to achieve up to 10 times faster and 14 times lower peak memory consumption compared to baseline implementations"](https://github.com/apple/ml-ane-transformers).

1. To utilize Core ML first, you need to convert a model from TensorFlow, PyTorch to Core ML model package format using [coremltools](https://coremltools.readme.io/docs) (or simply [utilize existing models in Core ML package format](https://github.com/likedan/Awesome-CoreML-Models) ).
2. Second, you must now use that converted package with an implementation designed for Apple Devices. Here is the [Apple XCode reference PyTorch implementation](https://github.com/apple/ml-ane-transformers).

https://machinelearning.apple.com/research/neural-engine-transformers

**Comment by vanhalt** [comment]
Hey there!

Sorry if I am misunderstanding the feature request.

How exactly would Enchanted interact with an interface that those models that take advantage of ANE & GPU? If I understand correctly this software only interacts thorough APIs with Ollama and OpenAI.

Am I missing something? Would love to understand!

**Comment by qdrddr** [comment]
CoreML is Apple's tech. You convert a model into CoreML format and it execute faster on Apple silicon. Like 3 times faster.

**Comment by bgiesing** [comment]
> CoreML is Apple's tech. You convert a model into CoreML format and it execute faster on Apple silicon. Like 3 times faster. 

That would be something you request Ollama to implement, not Enchanted. Enchanted is simply a native Mac GUI to Ollama which does all the actual model handeling, generation, etc. Ollama needs to support using CoreML.

**Comment by qdrddr** [comment]
Would that be of any help the [LM Studio has implemented MLX](https://github.com/lmstudio-ai/mlx-engine). And here is [Anemll ANE library](https://github.com/Anemll/Anemll) to work with MLX it is MIT Licensed.

**Comment by qdrddr** [comment]
there's [metal flash attention in swift by the way](https://github.com/philipturner/metal-flash-attention)

**Comment by qdrddr** [comment]
PoC of Metal Flash Attention with Python, C, Rust bindings for non-MLX models on Apple Silicon.

https://github.com/bghira/universal-metal-flash-attention
